Project skeleton

aleksandrkolesnikov opened the pull request.

Nipheris committed:
msixty - making your MSIX experience ten times better!

Nipheris committed:
initial CI setup

aleksandrkolesnikov committed:
Added lefthook

aleksandrkolesnikov committed:
Added conan

aleksandrkolesnikov committed:
Added src of msixty, conan recipe.

aleksandrkolesnikov committed:
Updated README.md

Nipheris commented on src/bootstrapper/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+)
+
+target_compile_definitions(${BOOTSTRAPPER_TARGET_NAME}
+    PRIVATE
```
Formatting (use tabs instead of spaces).

aleksandrkolesnikov replied:
Ok

Nipheris replied:
Or use 4 spaces in all the CMake scripts. At this moment formatting is inconsistent even within one file.

Nipheris commented on .clang-format:
```diff
@@ -0,0 +1,3 @@
+BasedOnStyle: Google
+IndentWidth: 4
+AccessModifierOffset: -4
```
Should we set tabs as an indentation method? AFAIK, Google mention spaces in their style guide.

aleksandrkolesnikov replied:
We don't use tabs for indentation. We use 4 spaces.

Nipheris commented on .circleci/config.yml:
```diff
@@ -0,0 +1,12 @@
+    steps:
+      - checkout
+      - run: pip install -r requirements.txt
+      - run: conan create . 0.0.0.0@tradingview/testing -s compiler.cppstd=17
```
We should use only semantic version as a Conan package version. I think `0.0.0` or `0.0.0-dev` should be enough. Later maybe we will implement a better computation logic for "development" version.

Nipheris commented on .gitignore:
```diff
@@ -0,0 +1,14 @@
+/.vs/*
+
+### Build artifacts ###
+out/
```
No newline at end of file).

aleksandrkolesnikov replied:
Ok, will add a newline

Nipheris commented on CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+include("${CMAKE_BINARY_DIR}/conanbuildinfo.cmake")
+conan_basic_setup()
+
+add_subdirectory(src)
```
Also newline.

Nipheris commented on CMakeSettings.json:
```diff
@@ -0,0 +1,40 @@
+      "variables": [
+        {
+          "name": "VERSION",
+          "value": "0.0.0.0",
```
Versions should also be fixed.

aleksandrkolesnikov replied:
Will do it

Nipheris commented on CMakeSettings.json:
```diff
@@ -0,0 +1,40 @@
+﻿{
+  "configurations": [
```
It seems this file is lost by Visual Studio autoformatting feature. Please, check it again in the other editor (VS Code for example).

aleksandrkolesnikov replied:
I did not understand. Which file? CMakeSettings.json?

Nipheris replied:
I mean we have two spaces in the beginning of the line before `"configurations"`, instead of tabs/four spaces/whatever we have in `tasks.vs.json`. I think we should keep one formatting style at least among JSON files.

aleksandrkolesnikov replied:
Actually, we have the same formatting in `tasks.vs.json`

aleksandrkolesnikov replied:
It seems github shows `tasks.vs.json` with incorrect indentations. If open this file in VS Code or Visual Studio the indentations are correct

Nipheris commented on README.md:
```diff
@@ -0,0 +1,29 @@
+
+This project uses:
+- [python 3.7](https://www.python.org/);
+- [`pipenv`](https://pipenv.pypa.io/) for working environment management;
```
Remove backticks here or add it for all three items - it looks strange now.

aleksandrkolesnikov replied:
Yeah, I will add them)

Nipheris commented on README.md:
````diff
@@ -0,0 +1,29 @@
+
+## Run conan for creating package:
+```
+$ pipenv run conan create . 1.2.3.4@tradingview/testing -s compiler.cppstd=17 --profile x86_64-Windows-v142-Debug -tbf out\\test_package_build\\x86_64-Debug
````
Use some semantic version as an example or replace it to something like `<package_version>` or even `0.0.0`. Also, the profile should not be used.

aleksandrkolesnikov replied:
Ok, got it

Nipheris commented on .vs/tasks.vs.json:
```diff
@@ -0,0 +1,65 @@
+				"PIPENV_NOSPIN": 1,
+				"PIPENV_VENV_IN_PROJECT": 1
+			},
+			"output": "HELL YEAH"
```
:D

aleksandrkolesnikov replied:
Will remove this)

Nipheris commented on .vs/tasks.vs.json:
```diff
@@ -0,0 +1,65 @@
+				"-s",
+				"compiler.cppstd=17",
+				"--profile",
+				"x86_64-Windows-v142-Debug"
```
We can not use this profile because it is not available for foreign developers. We should use command-line arguments or (possibly, better) commit this profiles in repo and use them in commands.

aleksandrkolesnikov replied:
Right, I will create profile. Should we support 32 bits platform?

Nipheris replied:
Yeah, because at least bootstrapper should be runnable on x86 (but it is not required for ALL binaries, for example CLI).

Nipheris commented on conanfile.py:
```diff
@@ -0,0 +1,49 @@
+
+class MsixtyConan(ConanFile):
+    name = "msixty"
+    license = "Proprietary"
```
We have chosen MIT license, see LICENSE file.

Nipheris commented on conanfile.py:
```diff
@@ -0,0 +1,49 @@
+        "src/*",
+        "CMakeLists.txt"
+    ]
+    generators = "cmake"
```
We should give a try to `CMakeToolchain` + `CMakeDeps`. I will open a separate issue after we merge this PR.

Nipheris commented on conanfile.py:
```diff
@@ -0,0 +1,49 @@
+    def package_info(self):
+        msixty_bin = os.path.join(self.package_folder, "bin")
+        self.output.info(f"Appending PATH environment variable: {msixty_bin}")
+        self.env_info.PATH.append(msixty_bin)
```
No newline.

Nipheris replied:
And do we really need this at all? `self.cpp_info.bindirs` should already do all the work.

aleksandrkolesnikov replied:
The default value of `self.cpp_info.bindirs` is `['bin']`. So, we still need to concatenate paths.

Nipheris commented on src/bootstrapper/main.cpp:
```diff
@@ -0,0 +1,57 @@
+
+class CMainWindow final : public CWindowImpl<CMainWindow, CWindow, CFrameWinTraits> {
+private:
+    // clang-format off
```
Tabs vs spaces.

Nipheris commented on src/cli/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+)
+
+install(TARGETS ${CLI_TARGET_NAME}
+	RUNTIME DESTINATION bin
```
AFAIK, `bin` directory is the default `RUNTIME` destination (and, anyway, this is configured by `cmake`/`CMakeToolchain`), so I think we don't need this line.

Nipheris commented on src/cli/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+
+target_compile_features(${CLI_TARGET_NAME}
+    PRIVATE
+        cxx_std_17
```
This should be configured by `cmake`/`CMakeToolchain`. We should investigate, it is it not.

Nipheris commented on src/bootstrapper/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+
+target_compile_features(${BOOTSTRAPPER_TARGET_NAME}
+	PRIVATE
+		cxx_std_17
```
This should be configured by `cmake`/`CMakeToolchain`. We should investigate, it is it not.

Nipheris commented on src/bootstrapper/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+)
+
+install(TARGETS ${BOOTSTRAPPER_TARGET_NAME}
+	RUNTIME DESTINATION bin
```
AFAIK, `bin` directory is the default `RUNTIME` destination (and, anyway, this is configured by `cmake`/`CMakeToolchain`), so I think we don't need this line.

Nipheris commented on src/bootstrapper/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+
+set_target_properties(${BOOTSTRAPPER_TARGET_NAME}
+	PROPERTIES
+		RUNTIME_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/${BOOTSTRAPPER_TARGET_NAME}/${CMAKE_BUILD_TYPE}"
```
Do we really need to reconfigure `RUNTIME_OUTPUT_DIRECTORY`?

Nipheris commented:
The simple `test_package` would be really helpful also.

Nipheris commented on .editorconfig:
```diff
@@ -0,0 +1,16 @@
+
+[*]
+charset = utf-8
+indent_style = tab
```
Let's configure spaces in `.editorconfig` also.

aleksandrkolesnikov committed:
Small refactoring.

- Added profiles for Conan
- Replaced commas with dots in *.rc.in.
- Unified indentations.

Nipheris committed:
migrate dependencies to pipenv
